File: tgboost/class_list.py
# design a ClassList,store (label,leaf node,prediction,grad,hess) ordered by data index

import numpy as np


class ClassList(object):
    def __init__(self, label):
        self.dataset_size = label.shape[0]
        self.label = label
        self.corresponding_tree_node = [None for _ in range(self.dataset_size)]
        self.pred = np.ones(label.shape, dtype="float32")
        self.grad = np.empty(label.shape, dtype="float32")
        self.hess = np.empty(label.shape, dtype="float32")

    def initialize_pred(self, first_round_pred):
        self.pred *= first_round_pred

    def update_pred(self, eta):
        for i in range(self.dataset_size):
            self.pred[i] += eta*self.corresponding_tree_node[i].leaf_score

    def update_grad_hess(self, loss):
        # self.grad = loss.grad(self.pred, self.label)
        # self.hess = loss.hess(self.pred, self.label)
        self.grad[:] = loss.grad(self.pred, self.label)
        self.hess[:] = loss.hess(self.pred, self.label)

    def sampling(self, row_mask):
        self.grad *= row_mask
        self.hess *= row_mask

    def statistic_given_inds(self, inds):
        # scan the given the index, calculate each alive tree node's (G,H)
        ret = {}
        for i in inds:
            tree_node = self.corresponding_tree_node[i]
            if tree_node.is_leaf:
                continue
            else:
                if tree_node.name not in ret:
                    ret[tree_node.name] = [0., 0.]
                ret[tree_node.name][0] += self.grad[i]
                ret[tree_node.name][1] += self.hess[i]
        return ret

    def update_corresponding_tree_node(self, treenode_leftinds_naninds):
        # scan the class list, if the data fall into tree_node
        # then we see whether its index is in left_inds or right_inds, update the corresponding tree node
        map = dict(treenode_leftinds_naninds)
        for i in range(self.dataset_size):
            tree_node = self.corresponding_tree_node[i]
            if not tree_node.is_leaf:
                left_inds = map[tree_node][0]
                nan_inds = map[tree_node][1]
                nan_go_to = map[tree_node][2]
                if i in left_inds:
                    self.corresponding_tree_node[i] = tree_node.left_child
                elif i in nan_inds:
                    if nan_go_to == 0:
                        self.corresponding_tree_node[i] = tree_node.nan_child
                    elif nan_go_to == 1:
                        self.corresponding_tree_node[i] = tree_node.left_child
                    else:
                        self.corresponding_tree_node[i] = tree_node.right_child
                else:
                    self.corresponding_tree_node[i] = tree_node.right_child

    def update_histogram_for_tree_node(self):
        # scan the class list
        # update histogram(Grad,Hess,num_sample) for each alive(new) tree node
        for i in range(self.dataset_size):
            tree_node = self.corresponding_tree_node[i]
            if not tree_node.is_leaf:
                tree_node.Grad_add(self.grad[i])
                tree_node.Hess_add(self.hess[i])
                tree_node.num_sample_add(1)

    def update_grad_hess_missing_for_tree_node(self, missing_value_attribute_list):
        for col in range(len(missing_value_attribute_list)):
            for i in missing_value_attribute_list[col]:
                tree_node = self.corresponding_tree_node[i]
                if not tree_node.is_leaf:
                    tree_node.Grad_missing[col] += self.grad[i]
                    tree_node.Hess_missing[col] += self.hess[i]

File: src/main/ClassList.java
package main;

import org.omg.CORBA.DATA_CONVERSION;

import java.util.ArrayList;
import java.util.Arrays;

public class ClassList {
    public int dataset_size;
    public double[] label;
    public TreeNode[] corresponding_tree_node;
    public double[] pred;
    public double[] grad;
    public double[] hess;

    public ClassList(TrainData data){
        this.dataset_size = data.dataset_size;
        this.label = data.label;
        this.pred = new double[dataset_size];
        this.grad = new double[dataset_size];
        this.hess = new double[dataset_size];
        this.corresponding_tree_node = new TreeNode[dataset_size];
    }

    public void initialize_pred(double first_round_pred){
        Arrays.fill(pred, first_round_pred);
    }

    public void update_pred(double eta){
        for(int i=0;i<dataset_size;i++){
            pred[i] += eta * corresponding_tree_node[i].leaf_score;
        }
    }

    public void update_grad_hess(Loss loss,double scale_pos_weight){
        grad = loss.grad(pred,label);
        hess = loss.hess(pred,label);
        if(scale_pos_weight != 1.0){
            for(int i=0;i<dataset_size;i++){
                if(label[i]==1){
                    grad[i] *= scale_pos_weight;
                    hess[i] *= scale_pos_weight;
                }
            }
        }
    }

    public void sampling(ArrayList<Double> row_mask){
        for(int i=0;i<dataset_size;i++){
            grad[i] *= row_mask.get(i);
            hess[i] *= row_mask.get(i);
        }
    }

    //TODO
    //parallel each col's calculation
    public void update_grad_hess_missing_for_tree_node(int[][] missing_value_attribute_list){
        for(int col=0;col<missing_value_attribute_list.length;col++){
            for(int i:missing_value_attribute_list[col]){
                TreeNode treenode = corresponding_tree_node[i];
                if(!treenode.is_leaf){
                    treenode.Grad_missing[col] += grad[i];
                    treenode.Hess_missing[col] += hess[i];
                }
            }
        }
    }

    public void update_Grad_Hess_numsample_for_tree_node(){
        for(int i=0;i<dataset_size;i++){
            TreeNode treenode = corresponding_tree_node[i];
            if(!treenode.is_leaf){
                treenode.Grad_add(grad[i]);
                treenode.Hess_add(hess[i]);
                treenode.num_sample_add(1);
            }
        }
    }

    public void update_corresponding_tree_node(AttributeList attribute_list){
        for(int i=0;i<dataset_size;i++){
            TreeNode treenode = corresponding_tree_node[i];
            if(!treenode.is_leaf){
                int split_feature = treenode.split_feature;
                double nan_go_to = treenode.nan_go_to;
                double val = attribute_list.origin_feature[i][split_feature];
                //consider categorical feature
                if(attribute_list.cat_features_cols.contains(split_feature)){
                    ArrayList<Double> left_child_catvalue = treenode.split_left_child_catvalue;
                    if(val==Data.NULL){
                        if(nan_go_to==0){
                            corresponding_tree_node[i] = treenode.nan_child;
                        }else if(nan_go_to==1){
                            corresponding_tree_node[i] = treenode.left_child;
                        }else {
                            corresponding_tree_node[i] = treenode.right_child;
                        }
                    }else if(left_child_catvalue.contains(val)){
                        corresponding_tree_node[i] = treenode.left_child;
                    }else {
                        corresponding_tree_node[i] = treenode.right_child;
                    }
                }else {
                    double split_threshold = treenode.split_threshold;
                    if(val== Data.NULL){
                        if(nan_go_to==0){
                            corresponding_tree_node[i] = treenode.nan_child;
                        }else if(nan_go_to==1){
                            corresponding_tree_node[i] = treenode.left_child;
                        }else {
                            corresponding_tree_node[i] = treenode.right_child;
                        }
                    }else if(val<=split_threshold){
                        corresponding_tree_node[i] = treenode.left_child;
                    }else {
                        corresponding_tree_node[i] = treenode.right_child;
                    }
                }
            }
        }
    }

}
